tests/integration_tests/functional/test_balloon.py

```diff
@@ -0,0 +1,713 @@
+# Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+"""Tests for guest-side operations on /balloon resources."""
+
+import logging
+import os
+import platform
+import subprocess
+import time
+
+import pytest
+
+from conftest import _test_images_s3_bucket
+from framework.artifacts import ArtifactCollection, ArtifactSet
+from framework.builder import MicrovmBuilder, SnapshotBuilder, SnapshotType
+from framework.matrix import TestMatrix, TestContext
+from framework.utils import get_free_mem_ssh
+
+import host_tools.network as net_tools  # pylint: disable=import-error
+
+
+MB_TO_PAGES = 256
+
+
+def get_rss_mem_by_pid(pid):
+    """Get the RSS memory that a guest uses, given the pid of the guest."""
+    output = subprocess.check_output("pmap -X {}".format(pid), shell=True)
+    return int(output.decode('utf-8').split('\n')[-2].split()[1], 10)
+
+
+def make_guest_dirty_memory(ssh_connection, should_oom=False, amount=8192):
+    """Tell the guest, over ssh, to dirty `amount` pages of memory."""
+    amount_in_mbytes = amount / MB_TO_PAGES
+
+    exit_code, _, _ = ssh_connection.execute_command(
+        "/sbin/fillmem {}".format(amount_in_mbytes)
+    )
+    time.sleep(0.1)
+
+    cmd = "cat /tmp/fillmem_output.txt"
+    _, stdout, _ = ssh_connection.execute_command(cmd)
+    if should_oom:
+        assert exit_code == 0 and (
+            "OOM Killer stopped the program with "
+            "signal 9, exit code 0"
+        ) in stdout.read()
+    else:
+        assert exit_code == 0 and (
+            "Memory filling was "
+            "successful"
+        ) in stdout.read()
+    time.sleep(0.1)
+
+
+def build_test_matrix(network_config, bin_cloner_path, logger):
+    """Build a test matrix using the kernel with the balloon driver."""
+    artifacts = ArtifactCollection(_test_images_s3_bucket())
+    # Testing matrix:
+    # - Guest kernel: Linux 4.14
+    # - Rootfs: Ubuntu 18.04
+    # - Microvm: 2vCPU with 256 MB RAM
+    # TODO: Multiple microvm sizes must be tested in the async pipeline.
+    microvm_artifacts = ArtifactSet(artifacts.microvms(keyword="2vcpu_256mb"))
+    kernel_artifacts = ArtifactSet(artifacts.kernels(
+        keyword="vmlinux-4.14"
+    ))
+    disk_artifacts = ArtifactSet(artifacts.disks(keyword="ubuntu"))
+
+    # Create a test context and add builder, logger, network.
+    test_context = TestContext()
+    test_context.custom = {
+        'builder': MicrovmBuilder(bin_cloner_path),
+        'network_config': network_config,
+        'logger': logger,
+        'snapshot_type': SnapshotType.FULL,
+        'seq_len': 5
+    }
+
+    # Create the test matrix.
+    return TestMatrix(
+        context=test_context,
+        artifact_sets=[
+            microvm_artifacts,
+            kernel_artifacts,
+            disk_artifacts
+        ]
+    )
+
+
+def create_vm(context, deflate=True, stats_interval=0):
```

Comment:
You can use VMMicro.spawn() to simplify this function.
Why are you enabling diff snapshots ?

Reply:
Also, I do not see where this fn is called from ?

Reply:
I removed it completely.